core/types/block_ext_test.go

```diff
@@ -0,0 +1,190 @@
+		extras.Header.Set(want, headerExtra)
+
+		assert.Equal(t, want, cpy)
+	})
+
+	t.Run("filled_header", func(t *testing.T) {
+		h := &Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		headerExtra := &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+		extras.Header.Set(h, headerExtra)
+
+		allFieldsAreSet(t, h)
+		allFieldsAreSet(t, headerExtra)
+
+		cpy := CopyHeader(h)
+
+		want := &Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		headerExtra = &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+		extras.Header.Set(want, headerExtra)
+
+		assert.Equal(t, want, cpy)
+
+		// Mutate each non-value field to ensure they are not shared
+		fieldsAreDeepCopied(t, h, cpy)
+		fieldsAreDeepCopied(t, GetHeaderExtra(h), GetHeaderExtra(cpy))
+	})
+}
+
+func ptrTo[T any](x T) *T { return &x }
+
+func allFieldsAreSet(t *testing.T, x any) {
+	t.Helper()
+	require.Equal(t, reflect.Ptr.String(), reflect.TypeOf(x).Kind().String(), "x must be a pointer")
+	v := reflect.ValueOf(x).Elem()
+	typ := v.Type()
+	require.Equal(t, reflect.Struct.String(), typ.Kind().String())
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldName := typ.Field(i).Name
+		fieldValue := field
+		if unicode.IsLower(rune(fieldName[0])) { // unexported
+			if field.Kind() == reflect.Ptr {
+				require.Falsef(t, field.IsNil(), "field %q is nil", fieldName)
+			}
+			field = reflect.NewAt(field.Type(), unsafe.Pointer(field.UnsafeAddr())).Elem() //nolint:gosec
```

Comment:
This trick is too obscure to be used without commentary (and IMO to be used at all outside of tests and even then only when absolutely necessary).

[In other words](https://tenor.com/en-GB/view/april-ludgate-thats-gross-i-love-it-parks-and-recreation-gif-11076075)

Reply:
Yeah kind of ugly way to check unexported fields. The latest code using your commit doesn't check unexported fields anymore, but I suppose that's fine for the time being. Should we however worry about future addition of unexported fields and us forgetting to deep copy them?

Reply:
TL;DR no

If something unexported is added by geth then it won't have a direct effect on RLP nor JSON so, from the perspective of a libevm consumer, it's not something that needs to be worried about. `CopyHeader` will use the upstream code anyway, so only the `HeaderExtra` copy needs to be tested.